sdk/core/core-rest-pipeline/src/util/tokenCycler.ts

```diff
@@ -52,50 +44,61 @@ export const DEFAULT_CYCLER_OPTIONS: TokenCyclerOptions = {
 };
 
 /**
- * Converts an an unreliable access token getter (which may resolve with null)
- * into an AccessTokenGetter by retrying the unreliable getter in a regular
- * interval.
+ * Calls to asyncFn until the timeout is reached.
+ * Prevents errors from bubbling up until the refresh timeout is reached.
+ * Once the timeout is reached, errors will bubble up, and null return values will throw with the received "errorMessage".
  *
- * @param getAccessToken - A function that produces a promise of an access token that may fail by returning null.
- * @param retryIntervalInMs - The time (in milliseconds) to wait between retry attempts.
- * @param refreshTimeout - The timestamp after which the refresh attempt will fail, throwing an exception.
- * @returns - A promise that, if it resolves, will resolve with an access token.
+ * @param asyncFn - The asynchronous function to call.
+ * @param timeout - Unix time when to stop ignoring errors or null values.
+ * @param errorMessage - Message to be used if the final value after the timeout is reached is still "null".
+ * @returns - A promise that can be resolved with T or null until the time is reached, after which it will throw if it can't resolve with a value of the type T.
  */
-async function beginRefresh(
-  getAccessToken: () => Promise<AccessToken | null>,
-  retryIntervalInMs: number,
-  refreshTimeout: number
-): Promise<AccessToken> {
-  // This wrapper handles exceptions gracefully as long as we haven't exceeded
-  // the timeout.
-  async function tryGetAccessToken(): Promise<AccessToken | null> {
-    if (Date.now() < refreshTimeout) {
-      try {
-        return await getAccessToken();
-      } catch {
-        return null;
-      }
-    } else {
-      const finalToken = await getAccessToken();
-
-      // Timeout is up, so throw if it's still null
-      if (finalToken === null) {
-        throw new Error("Failed to refresh access token.");
-      }
-
-      return finalToken;
+async function tryUntilTimeout<T>(
+  asyncFn: () => Promise<T | null>,
+  timeout: number,
+  errorMessage: string
+): Promise<T | null> {
+  if (Date.now() < timeout) {
+    try {
+      return asyncFn();
+    } catch {
+      return null;
+    }
+  } else {
+    const result = await asyncFn();
+
+    // Timeout is up, so throw if it's still null
+    if (result === null) {
+      throw new Error(errorMessage);
```

Comment:
Rather than passing `errorMessage` all the way through this chain of operations, we could throw a generic message here like `throw new Error("operation timed out")`, and catch/re-throw with more context information, such as `"failed to retrieve access_token: operation timed out"` when re-thrown from the token refresh function.